docs/installation.rst

```diff
@@ -408,3 +408,14 @@ needs to create a new Slack App and store its credentials on the Metagov server:
 
         systemctl reload apache2
 
+
+Github
+""""""
+
+In order to ues the Metagov Github plugin, the Metagov server administrator needs to create a new Github app and link it to Metagov:
+
+1. Create a metagov app and get the app ID. You can follow `this guide <https://docs.github.com/en/developers/apps/building-github-apps/creating-a-github-app>`_. Don't forget to `set permissions and subscribe to events <https://docs.github.com/en/developers/apps/managing-github-apps/editing-a-github-apps-permissions>`_.
+2. On Github, generate and download a `private key <https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#generating-a-private-key>`_. Put the private key in the github plugin folder.
+3. Make a copy of ``.env.example`` and rename it to ``.env``. Put the app ID in the file as well as the path to your private key.
+
+The plugin should now work. To use the app in their community, an admin will have to install the app manually on Github. They will then provide the installation ID and organization name as configuration parameters when enabling the plugin. We are working to make this process smoother in the future.
```

Comment:
Ok, I got this working (at least for creating actions, didn't test the webhook part). The GitHub app for the policykit.metagov.org instance–my dev instance– is https://github.com/apps/policykit-dev.

When I created the app in GitHub though, it asked me for a webhook URL and callback URL. If we change this to be similar to the Slack Plugin, those URLs would be `https://prototype.metagov.org/api/hooks/github` and `https://prototype.metagov.org/auth/github/callback`

In order to support oauth installation, the GitHub plugin needs to have a `views.py` file implementing `auth_callback` and `get_authorize_url` (see Slack examples). Then we'll be able to support a one-click installation for the GitHub plugin. That means the user wouldn't need to go manually install the app in GitHub.

and then, in order for GitHub to support a "global" webhook receiver, the plugin also needs to implement `process_event` in the views.

Reply:
here are some more details on how to implement support for oauth installation, in case you'll work on this next week when I'm off:
* In the github plugin, `views.get_authorize_url` should return `https://github.com/apps/<app name>/installations/new?state={state}`. (docs: https://docs.github.com/en/developers/apps/managing-github-apps/installing-github-apps#allowing-people-to-install-your-public-app-on-their-repository )
* When a driver navigates to `https://prototype.metagov.org/auth/github/authorize?communty=123&redirect_uri=xyz`, they are requesting to enable the GitHub plugin for the specified community. Metagov core will redirect them to the URL that you defined in get_authorize_url. (docs: https://prototype.metagov.org/redoc/#operation/Authorize%20plugin)
* When the use approves installation of the app, GitHub will redirect them to whatever URL is set as the Callback URL in the app setup. As mentioned, it should be `https://prototype.metagov.org/auth/github/callback`.
* When metagov core gets a request at `https://prototype.metagov.org/auth/github/callback` it will invoke the github plugin function `views.auth_callback(type, code, redirect_uri, community, state=None)`. In the GItHub plugin, we'd implement that function so that it creates a new GitHub plugin for the requested community, and ultimately returns to whatever redirect_uri was sent from the driver originally (the redirect URI from the orig request to https://prototype.metagov.org/redoc/#operation/Authorize%20plugin)
  * these auth functions were set up to support the slack installation and login flows, so it's possible it wont work out of the box for github. I think it should, though. For Slack, installation and user login use separate flows –– that's why we have `AuthType.APP_INSTALL` and `AuthType.USER_LOGIN`. I think you can just implement support for `AuthType.APP_INSTALL` for now, and ignore the user login type. GitHub also has support for asking for user identity _during_ installation, which slack does not.

Reply:
Thanks for this! Will likely take a stab at this during the week and we can go over it when you're back next week.

It seems like some of the instructions you're giving me on this PR should make their way into the docs, no?

Reply:
Okay @mashton ready for your review again when you're back from vacation. I implemented the generic endpoint and oauth logic, and also rewrote your advice here into the docs. My main confusion is with `auth_callback` -- what is `code` that's passed in to the function? Where can I access the information passed to me by Github? I'm not actually clear what Github's passing but it should have the installation_id and name of the owner it was installed to, that I can use to configure the plugin instance.

Reply:
> It seems like some of the instructions you're giving me on this PR should make their way into the docs, no?

Yes, thanks for adding it. Do you think the current implementation is..reasonable? I held off documenting these (`process_event` and oauth functions `get_authorize_url` and `auth_callback`) because they were my quick-and-dirty approach to get the Slack Plugin working quickly. I'd love to get your thoughts on how to improve this. There are a few things about it that are clunky to me:
* We're using import_module [here](https://github.com/metagov/metagov-prototype/blob/bef16576044e9aa56cedfe825ab6aeeb478ac9bc/metagov/metagov/core/views.py#L212) to load this file. There isn't an *interface* for the plugin developer to implement, they just have to know which function names to write and what to name the file. Maybe this is normal in Python (?) but it seems confusing to me and relies heavily on documentation. As a plugin author, I'd want some typed interface like "PluginAuth" that I can override/implement.
* The current approach to "global webhook receiver" requires the plugin author to "route" each request to each plugin and process, which means they need to understand the Django data model and use QuerySet filtering correctly. That's not ideal IMO. We could implement this as cleaner functional methods that plugin authors implement (see "webhook middleware chain" idea https://github.com/metagov/metagov-prototype/issues/99)

Reply:
> My main confusion is with auth_callback -- what is `code` that's passed in to the function?

The "code" is the [authorization code](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1). Both GitHub and Slack support the authorization code grant type. The code can be exchanged for an access token. You can read about that in the github docs [here](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#2-users-are-redirected-back-to-your-site-by-github).
BTW the params are described in the Slack plugin:
    https://github.com/metagov/metagov-prototype/blob/bef16576044e9aa56cedfe825ab6aeeb478ac9bc/metagov/metagov/plugins/slack/views.py#L69-L84

> Where can I access the information passed to me by Github?

Right now, you can't. GitHub passes some additional params that are not part of the oauth2 spec, so you'll need to pass them along to your auth_callback handler. The easiest way for you to get those is probably to have [core pass along](https://github.com/metagov/metagov-prototype/blob/master/metagov/metagov/core/views.py#L277-L283) any additional params that are present on `request.GET` using the object spread syntax (`**additional_parameters`).